components/script/dom/document.rs

```diff
@@ -514,19 +514,52 @@ impl Document {
 
     /// Attempt to find a named element in this page's document.
     /// https://html.spec.whatwg.org/multipage/#the-indicated-part-of-the-document
+    // 1. Apply the URL parser algorithm to the URL, and let fragid be the fragment component
+    //    of the resulting parsed URL.
     pub fn find_fragment_node(&self, fragid: &str) -> Option<Root<Element>> {
-        self.get_element_by_id(&Atom::from(fragid)).or_else(|| {
-            let check_anchor = |node: &HTMLAnchorElement| {
-                let elem = node.upcast::<Element>();
-                elem.get_attribute(&ns!(), &atom!("name"))
-                    .map_or(false, |attr| &**attr.value() == fragid)
-            };
-            let doc_node = self.upcast::<Node>();
-            doc_node.traverse_preorder()
-                    .filter_map(Root::downcast)
-                    .find(|node| check_anchor(&node))
-                    .map(Root::upcast)
-        })
+        // 2. If fragid is the empty string, then the indicated part of the document is the top
+        //    of the document; stop the algorithm here.
+        if fragid == "" {
+            self.GetDocumentElement()
+        } else {
+            use url::percent_encoding::percent_decode;
+            // 3. Let fragid bytes be the result of percent-decoding fragid.
```

Comment:
Step 2 is directly above, on line 520.
Step 1 is about getting the fragid from the url, but the value is already obtained elsewhere and passed in as an argument. I stuck that step above the function declaration to document that it arises elsewhere, since it wouldn't be very helpful at the actual point where it happens (in another file).